### ISSUE TITLE
[conflict] 'ssrf-proxy' and 'yawast'

### ISSUE THREAD

**Issue by jnbrains** [BlackArch/blackarch#1225]
Hi,

  I'm an archlinux user with blackarch repo enabled and most of the tools installed.
Todays update is adding new tools to my system, but I'm getting transaction
commit fail due to 'conflicting files'. 

  The packages in question are: '**ssrf-proxy**' and '**yawast**':

> $ pacaur -Si ssrf-proxy
> Repository      : blackarch
> Name            : ssrf-proxy
> Version         : 224.2e4c4a7-1
> Description     : Facilitates tunneling HTTP communications through servers vulnerable to Server-Side Request Forgery.
> Architecture    : any
> URL             : https://github.com/bcoles/ssrf_proxy
> Licenses        : MIT
> Groups          : blackarch  blackarch-proxy  blackarch-tunnel
> Provides        : None
> Depends On      : ruby  ruby-bundler
> Optional Deps   : None
> Conflicts With  : None
> Replaces        : None
> Download Size   : 943.43 KiB
> Installed Size  : 3474.00 KiB
> Packager        : Levon Kayan <noptrix@nullsecurity.net>
> Build Date      : Sun 06 Nov 2016 11:56:49 AM EET
> Validated By    : MD5 Sum  SHA-256 Sum  Signature
> 
> [lordy@beast: ~]
> $ pacaur -Si yawast
> Repository      : blackarch
> Name            : yawast
> Version         : 327.b6e39f8-1
> Description     : The YAWAST Antecedent Web Application Security Toolkit.
> Architecture    : x86_64
> URL             : https://github.com/adamcaudill/yawast
> Licenses        : BSD
> Groups          : blackarch  blackarch-webapp  blackarch-scanner  blackarch-fuzzer
> Provides        : None
> Depends On      : ruby
> Optional Deps   : None
> Conflicts With  : None
> Replaces        : None
> Download Size   : 2.34 MiB
> Installed Size  : 8.78 MiB
> Packager        : Levon Kayan <noptrix@nullsecurity.net>
> Build Date      : Sun 13 Nov 2016 02:53:21 PM EET
> Validated By    : MD5 Sum  SHA-256 Sum  Signature

  And the conflicting files:

> /usr/lib/ruby/gems/2.3.0/cache/colorize-0.8.1.gem exists in both 'ssrf-proxy' and 'yawast'
> /usr/lib/ruby/gems/2.3.0/cache/ipaddress-0.8.3.gem exists in both 'ssrf-proxy' and 'yawast'
> /usr/lib/ruby/gems/2.3.0/gems/colorize-0.8.1/CHANGELOG exists in both 'ssrf-proxy' and 'yawast'
> /usr/lib/ruby/gems/2.3.0/gems/colorize-0.8.1/LICENSE exists in both 'ssrf-proxy' and 'yawast'
> /usr/lib/ruby/gems/2.3.0/gems/colorize-0.8.1/README.md exists in both 'ssrf-proxy' and 'yawast'
> /usr/lib/ruby/gems/2.3.0/gems/colorize-0.8.1/Rakefile exists in both 'ssrf-proxy' and 'yawast'
> /usr/lib/ruby/gems/2.3.0/gems/colorize-0.8.1/colorize.gemspec exists in both 'ssrf-proxy' and 'yawast'
> /usr/lib/ruby/gems/2.3.0/gems/colorize-0.8.1/lib/colorize.rb exists in both 'ssrf-proxy' and 'yawast'
> /usr/lib/ruby/gems/2.3.0/gems/colorize-0.8.1/lib/colorize/class_methods.rb exists in both 'ssrf-proxy' and 'yawast'
> /usr/lib/ruby/gems/2.3.0/gems/colorize-0.8.1/lib/colorize/instance_methods.rb exists in both 'ssrf-proxy' and 'yawast'
> /usr/lib/ruby/gems/2.3.0/gems/colorize-0.8.1/lib/colorized_string.rb exists in both 'ssrf-proxy' and 'yawast'
> /usr/lib/ruby/gems/2.3.0/gems/colorize-0.8.1/test/test_colorize.rb exists in both 'ssrf-proxy' and 'yawast'
> /usr/lib/ruby/gems/2.3.0/gems/ipaddress-0.8.3/.document exists in both 'ssrf-proxy' and 'yawast'
> /usr/lib/ruby/gems/2.3.0/gems/ipaddress-0.8.3/.gitignore exists in both 'ssrf-proxy' and 'yawast'
> /usr/lib/ruby/gems/2.3.0/gems/ipaddress-0.8.3/.rock.yml exists in both 'ssrf-proxy' and 'yawast'
> /usr/lib/ruby/gems/2.3.0/gems/ipaddress-0.8.3/CHANGELOG.rdoc exists in both 'ssrf-proxy' and 'yawast'
> /usr/lib/ruby/gems/2.3.0/gems/ipaddress-0.8.3/Gemfile exists in both 'ssrf-proxy' and 'yawast'
> /usr/lib/ruby/gems/2.3.0/gems/ipaddress-0.8.3/LICENSE.txt exists in both 'ssrf-proxy' and 'yawast'
> /usr/lib/ruby/gems/2.3.0/gems/ipaddress-0.8.3/README.rdoc exists in both 'ssrf-proxy' and 'yawast'
> /usr/lib/ruby/gems/2.3.0/gems/ipaddress-0.8.3/Rakefile exists in both 'ssrf-proxy' and 'yawast'
> /usr/lib/ruby/gems/2.3.0/gems/ipaddress-0.8.3/ipaddress.gemspec exists in both 'ssrf-proxy' and 'yawast'
> /usr/lib/ruby/gems/2.3.0/gems/ipaddress-0.8.3/lib/ipaddress.rb exists in both 'ssrf-proxy' and 'yawast'
> /usr/lib/ruby/gems/2.3.0/gems/ipaddress-0.8.3/lib/ipaddress/ipv4.rb exists in both 'ssrf-proxy' and 'yawast'
> /usr/lib/ruby/gems/2.3.0/gems/ipaddress-0.8.3/lib/ipaddress/ipv6.rb exists in both 'ssrf-proxy' and 'yawast'
> /usr/lib/ruby/gems/2.3.0/gems/ipaddress-0.8.3/lib/ipaddress/mongoid.rb exists in both 'ssrf-proxy' and 'yawast'
> /usr/lib/ruby/gems/2.3.0/gems/ipaddress-0.8.3/lib/ipaddress/prefix.rb exists in both 'ssrf-proxy' and 'yawast'
> /usr/lib/ruby/gems/2.3.0/gems/ipaddress-0.8.3/lib/ipaddress/version.rb exists in both 'ssrf-proxy' and 'yawast'
> /usr/lib/ruby/gems/2.3.0/gems/ipaddress-0.8.3/test/ipaddress/ipv4_test.rb exists in both 'ssrf-proxy' and 'yawast'
> /usr/lib/ruby/gems/2.3.0/gems/ipaddress-0.8.3/test/ipaddress/ipv6_test.rb exists in both 'ssrf-proxy' and 'yawast'
> /usr/lib/ruby/gems/2.3.0/gems/ipaddress-0.8.3/test/ipaddress/mongoid_test.rb exists in both 'ssrf-proxy' and 'yawast'
> /usr/lib/ruby/gems/2.3.0/gems/ipaddress-0.8.3/test/ipaddress/prefix_test.rb exists in both 'ssrf-proxy' and 'yawast'
> /usr/lib/ruby/gems/2.3.0/gems/ipaddress-0.8.3/test/ipaddress_test.rb exists in both 'ssrf-proxy' and 'yawast'
> /usr/lib/ruby/gems/2.3.0/gems/ipaddress-0.8.3/test/test_helper.rb exists in both 'ssrf-proxy' and 'yawast'
> /usr/lib/ruby/gems/2.3.0/specifications/colorize-0.8.1.gemspec exists in both 'ssrf-proxy' and 'yawast'
> /usr/lib/ruby/gems/2.3.0/specifications/ipaddress-0.8.3.gemspec exists in both 'ssrf-proxy' and 'yawast'

Regards,
JNB

**Comment by noptrix** [comment]
just use the --force option here for pacman.

**Comment by jnbrains** [comment]
I'm aware of the --force option. It's just that that way it's going to mess the system up.
I believe there should be a cleaner solution for problems like this, not to simply force it.

Thanks anyway.

**Comment by lsfxz** [comment]
If the blackarch devs would be interested in this, I'd be willing to try and "fix" this, as in: create "proper" packages for all the ruby-blackarch-applications and their requirements (like it's done with the quarry repo, for example). 
I've already done that for the packages that gave me conflicts, though I'd probably try to automate this somewhat if I'd do it on a larger scale.

**Comment by noptrix** [comment]
that would rock, @lsfxz. go ahead :)

**Comment by lsfxz** [comment]
Great! 
I just realized the quarry repo is on github and could be utilized to automate a huge part of that process. I'll get to it asap (probably on monday) and get back here when done :)

**Comment by noptrix** [comment]
@lsfxz  perfect, thank you!

**Comment by lsfxz** [comment]
Ok, so:
Here's the result: https://github.com/lsfxz/blackarch/tree/fix-ruby
(I modified quarry: https://github.com/lsfxz/quarry/tree/blackarch and did some manual fine-tuning, resulting in these pkgbuilds: https://github.com/lsfxz/blackarch-ruby )

I did it as "Arch" as possible, so now all ruby packages have a "ruby-" prefix and their dependencies are also provided as packages. The newly-prefixed packages replace, conflict and, for backwards compatibility, provide their original counterparts.

Unfortunately from a quick look at the blackarch build process I couldn't easily test if there were issues there, and also couldn't test if an update (-Syu) would correctly update and replace the packages: I've only tested it with a local repository on a blackarch VM: There the manual installation and replacement worked, and the few randomly picked packages I tested it with worked flawlessly when called afterwards. The automatic update didn't work, though.

I've also tried it on x86_64 only: It _should_ work on x86_32 and ARM but I don't know if it does.
